### PR TITLE
Use consistent version comparison even when comparing preferred versions

### PR DESCRIPTION
**Pull request by big-guy**

### CODE REVIEW

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/selectors/SelectorStateResolver.java [hunk 1]
```diff
@@ -169,7 +169,7 @@ private TreeSet<ComponentIdResolveResult> maybeResolvePreferConstraint(TreeSet<C
         ComponentIdResolveResult resolvedPreference = selector.resolvePrefer(allRejects);
         if (resolvedPreference != null && resolvedPreference.getFailure() == null) {
             if (preferResults == null) {
-                preferResults = Sets.newTreeSet(new DescendingResolveResultComparator());
+                preferResults = Sets.newTreeSet(new DescendingResolveResultComparator(versionComparator, versionParser));
```
Comment (big-guy) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/selectors/SelectorStateResolver.java:172]: @ljacomet any idea if we intended preferred versions to use a different comparison scheme?

The scenario here:

We have a dependency on an external module and we need to select between versions `1.0-release` and `1.0-releaseProguard`. 

If they're required versions, we select `1.0-release`.
If they're preferred versions, we select `1.0-releaseProguard`.
If they're preferred versions and there's a dynamic required version anywhere, we select `1.0-release`. 

With this change, we select `1.0-release` in all cases.

**Comment by big-guy** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/selectors/SelectorStateResolver.java (hunk 1 above)
Comment (ljacomet) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/selectors/SelectorStateResolver.java:172]: This looks like a bug left when we reworked version comparison.
I don't see what would be the use case to have a slightly different compare in some cases.

👍 from me

**Comment by big-guy** [comment]
fixes #27498

**Pull request closed without merging**